common/kvstore/tablestore/table_store_builder.go

```diff
@@ -0,0 +1,374 @@
+// This is required for atomic deletion of tables.
+const metadataDeletionKey = "deletion"
+
+// The current schema version of the metadata table.
+const currentSchemaVersion uint64 = 0
+
+// StoreType describes the underlying store implementation.
+type StoreType int
+
+const (
+	// LevelDB is a LevelDB-backed store.
+	LevelDB StoreType = iota
+	// MapStore is an in-memory store. This store does not preserve data across restarts.
+	MapStore
+)
+
+// Create creates a new TableStore instance of the given type. The store will be created at the given path.
+// This method will set up a table for each table name provided, and will drop all tables not in the list.
+// Dropping a table is irreversible and will delete all data in the table, so be very careful not to call
+// this method with table names omitted by mistake.
+func (t StoreType) Create(logger logging.Logger, path string, tables ...string) (kvstore.TableStore, error) {
+
+	var base kvstore.Store
+	var err error
+
+	switch t {
+	case LevelDB:
+		base, err = leveldb.NewStore(logger, path)
+		if err != nil {
+			return nil, fmt.Errorf("error creating LevelDB store: %w", err)
+		}
+	case MapStore:
+		base = mapstore.NewStore()
+	default:
+		return nil, fmt.Errorf("unknown store type: %d", t)
+	}
+
+	return create(logger, base, tables...)
+}
+
+// Future work: if we ever decide to permit third parties to provide custom store implementations not in this module,
+// we will need to make this method public.
+
+// create creates a new TableStore instance with the given base store and table names.
+func create(logger logging.Logger, base kvstore.Store, tables ...string) (kvstore.TableStore, error) {
+	metadataTable := newTableView(base, "metadata", metadataTableID)
+	namespaceTable := newTableView(base, "namespace", namespaceTableID)
+
+	err := validateSchema(metadataTable)
+	if err != nil {
+		return nil, fmt.Errorf("error validating schema: %w", err)
+	}
+
+	tableIDMap, err := loadNamespaceTable(namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error loading namespace table: %w", err)
+	}
+
+	err = handleIncompleteDeletion(logger, base, metadataTable, namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error handling incomplete deletion: %w", err)
+	}
+
+	err = addAndRemoveTables(base, metadataTable, namespaceTable, tableIDMap, tables)
+	if err != nil {
+		return nil, fmt.Errorf("error adding and removing tables: %w", err)
+	}
+
+	tableMap := make(map[string]kvstore.Table, len(tableIDMap))
+	for tableID, tableName := range tableIDMap {
+		tableMap[tableName] = newTableView(base, tableName, tableID)
+	}
+	return newTableStore(logger, base, tableMap), nil
+}
+
+// validateSchema loads/initiates the schema version in the metadata table.
+func validateSchema(metadataTable kvstore.Table) error {
+
+	schemaKey := []byte(metadataSchemaVersionKey)
+	onDiskSchemaBytes, err := metadataTable.Get(schemaKey)
+
+	if errors.Is(err, kvstore.ErrNotFound) {
```

Comment:
The control below can be made more readable:
```
if err != nil {
    if !errors.Is(err, kvstore.ErrNotFound) {
        return ...
    }
    // handle not found case
    ...

   return nil
}

// Verify schema version
...
```

Reply:
refactored